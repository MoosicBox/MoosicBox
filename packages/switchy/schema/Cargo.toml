[package]
authors     = { workspace = true }
categories  = ["database", "development-tools"]
description = "Generic schema migration system for switchy_database"
edition     = { workspace = true }
keywords    = ["database", "migration", "schema", "switchy"]
license     = { workspace = true }
name        = "switchy_schema"
readme      = "README.md"
repository  = { workspace = true }
version     = "0.1.0"

[package.metadata.workspaces]
group = "switchy"

[dependencies]
async-trait      = { workspace = true }
switchy_database = { workspace = true, features = ["schema"] }
thiserror        = { workspace = true }

# Optional dependencies for discovery features
bytes = { workspace = true }
include_dir = { workspace = true, optional = true }
switchy_fs = { workspace = true, optional = true, features = [
    "async",
    "tokio",
] }

[dev-dependencies]
switchy_async = { workspace = true, features = ["macros", "tokio"] }
switchy_fs    = { workspace = true, features = ["simulator-real-fs", "std"] }
tokio         = { workspace = true, features = ["macros", "rt"] }

[features]
default = ["all-discovery"]

fail-on-warnings = []

all-discovery = ["code", "directory", "embedded"]
code          = []
directory     = ["dep:switchy_fs"]
embedded      = ["dep:include_dir"]
test-utils    = []
validation    = []
