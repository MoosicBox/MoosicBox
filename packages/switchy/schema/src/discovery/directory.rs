//! # Directory-Based Migrations
//!
//! This module provides support for directory-based migrations that are loaded
//! from the filesystem at runtime. This is useful for development environments
//! where you want to modify migrations without recompiling.
//!
//! ## Features
//!
//! * **Runtime loading**: Migrations loaded from disk when needed
//! * **Development friendly**: Modify migrations without rebuilding
//! * **Same structure**: Uses identical directory structure to embedded migrations
//! * **Hot reload**: Changes are picked up on next migration run
//! * **Optional files**: Both up.sql and down.sql are optional
//!
//! ## Directory Structure
//!
//! Directory-based migrations use the same structure as embedded migrations:
//!
//! ```text
//! migrations/
//! ├── 001_create_users/
//! │   ├── up.sql      # Optional migration SQL
//! │   └── down.sql    # Optional rollback SQL
//! ├── 002_add_posts/
//! │   └── up.sql      # down.sql is optional
//! └── 003_indexes/
//!     ├── up.sql
//!     └── down.sql
//! ```
//!
//! ## Usage
//!
//! ```rust,no_run
//! use switchy_schema::runner::MigrationRunner;
//!
//! # async fn example(db: &dyn switchy_database::Database) -> switchy_schema::Result<()> {
//! // Create migration runner for directory-based migrations
//! let runner = MigrationRunner::new_directory("./migrations");
//!
//! // Run migrations
//! runner.run(db).await?;
//! # Ok(())
//! # }
//! ```
//!
//! ## Advantages vs Embedded
//!
//! * **Development speed**: No rebuild required to change migrations
//! * **Debugging**: Easier to inspect and modify SQL during development
//! * **Flexibility**: Can load migrations from different paths
//!
//! ## Disadvantages vs Embedded
//!
//! * **Runtime dependency**: Requires filesystem access
//! * **Deployment complexity**: Must ensure migration files are available
//! * **Potential inconsistency**: Files can be modified or missing at runtime

use std::{collections::BTreeMap, path::PathBuf, sync::Arc};

use async_trait::async_trait;
use sha2::{Digest, Sha256};

use crate::{Result, migration::Migration, migration::MigrationSource};

/// A single file-based migration with optional up and down SQL content
///
/// File-based migrations are loaded from the filesystem at runtime,
/// providing flexibility for development and debugging. Each migration
/// consists of:
///
/// * **ID**: The directory name (used for ordering)
/// * **Path**: The filesystem path to the migration directory  
/// * **Up SQL**: Optional SQL for applying the migration
/// * **Down SQL**: Optional SQL for rolling back the migration
///
/// Both up and down SQL are optional. Missing or empty SQL files
/// are treated as no-op operations.
pub struct FileMigration {
    id: String,
    path: PathBuf,
    up_sql: Option<String>,
    down_sql: Option<String>,
}

impl FileMigration {
    #[must_use]
    pub const fn new(
        id: String,
        path: PathBuf,
        up_sql: Option<String>,
        down_sql: Option<String>,
    ) -> Self {
        Self {
            id,
            path,
            up_sql,
            down_sql,
        }
    }
    #[must_use]
    pub const fn path(&self) -> &PathBuf {
        &self.path
    }
}

#[async_trait]
impl Migration<'static> for FileMigration {
    fn id(&self) -> &str {
        &self.id
    }

    async fn up(&self, db: &dyn switchy_database::Database) -> Result<()> {
        if let Some(up_sql) = &self.up_sql
            && !up_sql.trim().is_empty()
        {
            db.exec_raw(up_sql).await?;
        }
        Ok(())
    }

    async fn down(&self, db: &dyn switchy_database::Database) -> Result<()> {
        if let Some(down_sql) = &self.down_sql
            && !down_sql.trim().is_empty()
        {
            db.exec_raw(down_sql).await?;
        }
        Ok(())
    }

    async fn up_checksum(&self) -> Result<bytes::Bytes> {
        let mut hasher = Sha256::new();
        match &self.up_sql {
            Some(content) => hasher.update(content.as_bytes()),
            None => hasher.update(b""), // Hash empty bytes for None
        }
        Ok(bytes::Bytes::from(hasher.finalize().to_vec()))
    }

    async fn down_checksum(&self) -> Result<bytes::Bytes> {
        let mut hasher = Sha256::new();
        match &self.down_sql {
            Some(content) => hasher.update(content.as_bytes()),
            None => hasher.update(b""), // Hash empty bytes for None
        }
        Ok(bytes::Bytes::from(hasher.finalize().to_vec()))
    }
}

/// Migration source for directory-based migrations loaded from the filesystem
///
/// This source loads migrations from a directory on the filesystem at runtime.
/// It's particularly useful during development when you want to modify migrations
/// without recompiling your application.
///
/// ## Features
///
/// * **Runtime loading**: Scans filesystem when migrations are requested
/// * **Development friendly**: Changes are picked up without rebuilds
/// * **Error handling**: Gracefully handles missing or unreadable files
/// * **Consistent ordering**: Migrations sorted alphabetically by directory name
///
/// ## Example
///
/// ```rust,no_run
/// use switchy_schema::{
///     runner::MigrationRunner,
///     discovery::directory::DirectoryMigrationSource
/// };
/// use std::path::PathBuf;
///
/// // Create source directly (advanced usage)
/// let source = DirectoryMigrationSource::from_path(PathBuf::from("./migrations"));
///
/// // Or use the convenience constructor (recommended)
/// let runner = MigrationRunner::new_directory("./migrations");
/// ```
///
/// The source will scan the specified directory for subdirectories containing
/// `up.sql` and/or `down.sql` files. Each subdirectory becomes a migration
/// with its name as the migration ID.
///
/// ## Error Handling
///
/// The directory source handles several filesystem-related scenarios:
///
/// * **Missing directory**: Returns empty migration list
/// * **Unreadable files**: Skips files that cannot be read
/// * **Empty files**: Treats as no-op migrations
/// * **Missing SQL files**: Creates migration with no-op for missing files
pub struct DirectoryMigrationSource {
    migrations_path: PathBuf,
}

impl DirectoryMigrationSource {
    #[must_use]
    pub const fn from_path(migrations_path: PathBuf) -> Self {
        Self { migrations_path }
    }

    #[must_use]
    pub const fn path(&self) -> &PathBuf {
        &self.migrations_path
    }
}

#[async_trait]
impl MigrationSource<'static> for DirectoryMigrationSource {
    async fn migrations(&self) -> Result<Vec<Arc<dyn Migration<'static> + 'static>>> {
        let migration_map = self.extract_migrations().await?;
        let migrations: Vec<Arc<dyn Migration<'static> + 'static>> = migration_map
            .into_values()
            .map(|m| Arc::new(m) as Arc<dyn Migration<'static> + 'static>)
            .collect();
        Ok(migrations)
    }
}

impl DirectoryMigrationSource {
    /// Extract migrations from the directory structure
    ///
    /// # Errors
    ///
    /// * If the migrations directory cannot be read
    /// * If any migration directory cannot be accessed
    /// * If any SQL file cannot be read
    async fn extract_migrations(&self) -> Result<BTreeMap<String, FileMigration>> {
        let mut migrations = BTreeMap::new();

        // Read all entries in the migrations directory
        let entries = switchy_fs::unsync::read_dir_sorted(&self.migrations_path).await?;

        for entry in entries {
            let entry_path = entry.path();

            // Only process directories (each directory is one migration)
            if !entry.file_type().await?.is_dir() {
                continue;
            }

            // Use directory name as migration ID (as-is, no validation)
            let migration_id = entry.file_name().to_string_lossy().to_string();

            // Look for up.sql and down.sql files
            let up_sql_path = entry_path.join("up.sql");
            let down_sql_path = entry_path.join("down.sql");

            // Read up.sql (optional)
            let up_sql = switchy_fs::unsync::read_to_string(&up_sql_path).await.ok();

            // Read down.sql (optional)
            let down_sql = switchy_fs::unsync::read_to_string(&down_sql_path)
                .await
                .ok();

            // Skip migrations with no SQL files at all
            if up_sql.is_none() && down_sql.is_none() {
                continue;
            }

            let migration = FileMigration::new(migration_id.clone(), entry_path, up_sql, down_sql);

            migrations.insert(migration_id, migration);
        }

        Ok(migrations)
    }
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use super::*;

    #[switchy_async::test(real_fs)]
    async fn test_directory_migration_source() {
        let test_dir = Path::new(env!("CARGO_MANIFEST_DIR")).join("test_migrations_dir");

        let source = DirectoryMigrationSource::from_path(test_dir);
        let migrations = source.migrations().await.unwrap();

        // Should find 3 migrations in alphabetical order (004_no_sql_files is skipped)
        assert_eq!(migrations.len(), 3);

        // Check migration IDs are in alphabetical order
        assert_eq!(migrations[0].id(), "001_create_users");
        assert_eq!(migrations[1].id(), "002_add_indexes");
        assert_eq!(migrations[2].id(), "003_empty_migration");
    }

    #[switchy_async::test(real_fs)]
    async fn test_file_migration_with_content() {
        let test_dir = Path::new(env!("CARGO_MANIFEST_DIR")).join("test_migrations_dir");

        let source = DirectoryMigrationSource::from_path(test_dir);
        let migration_map = source.extract_migrations().await.unwrap();

        // Test migration with both up and down SQL
        let create_users = migration_map.get("001_create_users").unwrap();
        assert!(
            create_users
                .up_sql
                .as_ref()
                .unwrap()
                .contains("CREATE TABLE users")
        );
        assert!(
            create_users
                .down_sql
                .as_ref()
                .unwrap()
                .contains("DROP TABLE users")
        );

        // Test migration with only up SQL
        let add_indexes = migration_map.get("002_add_indexes").unwrap();
        assert!(
            add_indexes
                .up_sql
                .as_ref()
                .unwrap()
                .contains("CREATE INDEX")
        );
        assert!(add_indexes.down_sql.is_none());

        // Test empty migration
        let empty_migration = migration_map.get("003_empty_migration").unwrap();
        assert!(empty_migration.up_sql.as_ref().unwrap().trim().is_empty());
        assert!(empty_migration.down_sql.is_none());

        // Test that migration with no SQL files is not included
        assert!(!migration_map.contains_key("004_no_sql_files"));
    }

    #[switchy_async::test]
    async fn test_file_migration_execution() {
        let migration = FileMigration::new(
            "test".to_string(),
            PathBuf::from("/test"),
            Some("SELECT 1;".to_string()),
            Some("SELECT 2;".to_string()),
        );

        // Test that migration has correct ID
        assert_eq!(migration.id(), "test");

        // Test that path is accessible
        assert_eq!(migration.path(), &PathBuf::from("/test"));
    }

    #[switchy_async::test]
    async fn test_empty_sql_handling() {
        let migration = FileMigration::new(
            "empty".to_string(),
            PathBuf::from("/empty"),
            Some(String::new()),
            Some("   ".to_string()),
        );

        // Empty SQL should not cause errors (tested via compilation)
        assert_eq!(migration.id(), "empty");
    }

    #[switchy_async::test]
    async fn test_migration_with_no_sql_files() {
        let migration =
            FileMigration::new("no_sql".to_string(), PathBuf::from("/no_sql"), None, None);

        // Migration with no SQL files should be valid but do nothing
        assert_eq!(migration.id(), "no_sql");
    }
}
