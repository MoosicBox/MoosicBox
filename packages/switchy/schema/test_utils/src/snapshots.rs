//! Snapshot testing utilities for migration verification using JSON format
//!
//! This module provides utilities for capturing and comparing database schemas
//! and migration results using insta's snapshot testing with JSON serialization.
//! JSON is used for its wide compatibility, active maintenance, and human readability
//! when pretty-printed.

use crate::TestError;
use std::path::PathBuf;
use switchy_database::{Database, DatabaseError};
use switchy_schema::MigrationError;

#[cfg(feature = "snapshots")]
use insta::Settings;
#[cfg(feature = "snapshots")]
use serde::{Deserialize, Serialize};
#[cfg(feature = "snapshots")]
use std::collections::BTreeMap;
#[cfg(feature = "snapshots")]
use std::sync::Arc;
#[cfg(feature = "snapshots")]
use std::{future::Future, pin::Pin};

#[cfg(feature = "snapshots")]
use switchy_database::schema::{ColumnInfo as DbColumnInfo, TableInfo};
#[cfg(feature = "snapshots")]
use switchy_database::{DatabaseValue, Row};
#[cfg(feature = "snapshots")]
use switchy_schema::discovery::directory::DirectoryMigrationSource;
#[cfg(feature = "snapshots")]
use switchy_schema::migration::{Migration, MigrationSource};
#[cfg(feature = "snapshots")]
use switchy_schema::runner::MigrationRunner;

#[cfg(feature = "snapshots")]
type SetupFn = Box<
    dyn for<'a> Fn(
            &'a dyn Database,
        ) -> Pin<
            Box<dyn Future<Output = std::result::Result<(), DatabaseError>> + Send + 'a>,
        > + Send
        + Sync,
>;
#[cfg(feature = "snapshots")]
type VerificationFn = Box<
    dyn for<'a> Fn(
            &'a dyn Database,
        ) -> Pin<
            Box<dyn Future<Output = std::result::Result<(), DatabaseError>> + Send + 'a>,
        > + Send
        + Sync,
>;

// VecMigrationSource helper for test utilities
#[cfg(feature = "snapshots")]
struct VecMigrationSource<'a> {
    migrations: Vec<Arc<dyn Migration<'a> + 'a>>,
}

#[cfg(feature = "snapshots")]
impl<'a> VecMigrationSource<'a> {
    fn new(migrations: Vec<Arc<dyn Migration<'a> + 'a>>) -> Self {
        Self { migrations }
    }
}

#[cfg(feature = "snapshots")]
#[async_trait::async_trait]
impl<'a> MigrationSource<'a> for VecMigrationSource<'a> {
    async fn migrations(&self) -> switchy_schema::Result<Vec<Arc<dyn Migration<'a> + 'a>>> {
        Ok(self.migrations.clone())
    }
}

/// Error type for snapshot testing operations
#[derive(Debug, thiserror::Error)]
pub enum SnapshotError {
    /// Database operation failed
    #[error("Database error: {0}")]
    Database(#[from] DatabaseError),

    /// Migration operation failed
    #[error("Migration error: {0}")]
    Migration(#[from] MigrationError),

    /// IO operation failed
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),

    /// Snapshot validation failed
    #[error("Snapshot validation failed: {0}")]
    Validation(String),

    /// Test utilities error
    #[error("Test error: {0}")]
    Test(#[from] TestError),

    /// JSON serialization/deserialization error
    #[error("JSON error: {0}")]
    Json(#[from] serde_json::Error),
}

/// Result type for snapshot operations
pub type Result<T> = std::result::Result<T, SnapshotError>;

/// Snapshot structure for migration state verification
/// Note: This structure will grow in later phases.
/// Breaking changes to snapshot structure are acceptable during development.
/// Regenerate snapshots with `cargo insta review` when structure changes.
#[cfg(feature = "snapshots")]
#[derive(Debug, Serialize, Deserialize)]
struct MigrationSnapshot {
    test_name: String,
    migration_sequence: Vec<String>,
    schema: Option<DatabaseSchema>,
    data_samples: Option<std::collections::BTreeMap<String, Vec<serde_json::Value>>>,
}

/// Complete database schema structure for snapshot storage
#[cfg(feature = "snapshots")]
#[derive(Debug, Serialize, Deserialize)]
struct DatabaseSchema {
    tables: BTreeMap<String, TableSchema>,
}

/// Table schema structure with columns and indexes
#[cfg(feature = "snapshots")]
#[derive(Debug, Serialize, Deserialize)]
struct TableSchema {
    columns: Vec<ColumnInfo>,
    indexes: Vec<String>,
}

/// Column information for schema capture
#[cfg(feature = "snapshots")]
#[derive(Debug, Serialize, Deserialize)]
struct ColumnInfo {
    name: String,
    data_type: String,
    nullable: bool,
    default_value: Option<String>,
    primary_key: bool,
}

/// Placeholder for snapshot testing functionality
/// Full implementation will come in Phase 11.4.2+
pub struct SnapshotTester {
    // Implementation to follow in subsequent phases
}

/// Migration snapshot test struct for verifying database schema changes
#[allow(clippy::struct_excessive_bools)]
pub struct MigrationSnapshotTest {
    test_name: String,
    migrations_dir: Option<PathBuf>,
    assert_schema: bool,
    assert_sequence: bool,
    expected_tables: Vec<String>, // NEW: Tables to inspect for schema capture
    redact_timestamps: bool,
    redact_auto_ids: bool,
    redact_paths: bool,
    assert_data: bool,
    data_samples: std::collections::BTreeMap<String, usize>,
    setup_fn: Option<SetupFn>,
    verification_fn: Option<VerificationFn>,
    db: Option<Box<dyn Database>>,
    migrations_table_name: Option<String>,
}

impl MigrationSnapshotTest {
    /// Create a new migration snapshot test
    #[must_use]
    pub fn new(test_name: &str) -> Self {
        Self {
            test_name: test_name.to_string(),
            // No default migrations directory
            migrations_dir: None,
            assert_schema: true,
            assert_sequence: true,
            expected_tables: Vec::new(), // Empty by default
            redact_timestamps: true,
            redact_auto_ids: true,
            redact_paths: true,
            assert_data: false,
            data_samples: std::collections::BTreeMap::new(),
            setup_fn: None,
            verification_fn: None,
            db: None,
            migrations_table_name: None,
        }
    }

    #[must_use]
    pub fn migrations_dir(mut self, path: impl Into<PathBuf>) -> Self {
        self.migrations_dir = Some(path.into());
        self
    }

    #[must_use]
    pub const fn assert_schema(mut self, enabled: bool) -> Self {
        self.assert_schema = enabled;
        self
    }

    #[must_use]
    pub const fn assert_sequence(mut self, enabled: bool) -> Self {
        self.assert_sequence = enabled;
        self
    }

    /// Configure which tables to inspect for schema capture
    #[must_use]
    pub fn expected_tables(mut self, tables: Vec<String>) -> Self {
        self.expected_tables = tables;
        self
    }

    /// Configure timestamp redaction
    #[must_use]
    pub const fn redact_timestamps(mut self, enabled: bool) -> Self {
        self.redact_timestamps = enabled;
        self
    }

    /// Configure auto-ID redaction
    #[must_use]
    pub const fn redact_auto_ids(mut self, enabled: bool) -> Self {
        self.redact_auto_ids = enabled;
        self
    }

    /// Configure path redaction
    #[must_use]
    pub const fn redact_paths(mut self, enabled: bool) -> Self {
        self.redact_paths = enabled;
        self
    }

    /// Configure data assertion
    #[must_use]
    pub const fn assert_data(mut self, enabled: bool) -> Self {
        self.assert_data = enabled;
        self
    }

    /// Add data samples for a specific table
    #[must_use]
    pub fn with_data_samples(mut self, table: &str, count: usize) -> Self {
        self.data_samples.insert(table.to_string(), count);
        self
    }

    /// Add setup function to run before migrations
    #[must_use]
    #[cfg(feature = "snapshots")]
    pub fn with_setup<F>(mut self, f: F) -> Self
    where
        F: for<'a> Fn(
                &'a dyn Database,
            ) -> Pin<
                Box<dyn Future<Output = std::result::Result<(), DatabaseError>> + Send + 'a>,
            > + Send
            + Sync
            + 'static,
    {
        self.setup_fn = Some(Box::new(f));
        self
    }

    /// Add verification function to run after migrations
    #[must_use]
    #[cfg(feature = "snapshots")]
    pub fn with_verification<F>(mut self, f: F) -> Self
    where
        F: for<'a> Fn(
                &'a dyn Database,
            ) -> Pin<
                Box<dyn Future<Output = std::result::Result<(), DatabaseError>> + Send + 'a>,
            > + Send
            + Sync
            + 'static,
    {
        self.verification_fn = Some(Box::new(f));
        self
    }

    /// Use an existing database instance instead of creating a new one
    /// This allows integration with `MigrationTestBuilder` or other test scenarios
    #[must_use]
    pub fn with_database(mut self, db: Box<dyn Database>) -> Self {
        self.db = Some(db);
        self
    }

    /// Set custom migrations table name (defaults to `__switchy_migrations`)
    #[must_use]
    pub fn with_migrations_table(mut self, table_name: impl Into<String>) -> Self {
        self.migrations_table_name = Some(table_name.into());
        self
    }

    /// Auto-discover tables by parsing migration files (future enhancement)
    #[must_use]
    pub const fn auto_discover_tables(self) -> Self {
        // Will be implemented to parse CREATE TABLE from migration files
        self
    }

    /// Full integration with existing test builder for complex scenarios
    #[must_use]
    #[cfg(feature = "snapshots")]
    pub fn with_test_builder(self, _builder: crate::MigrationTestBuilder<'_>) -> Self {
        // Integration bridges the two systems
        // Note: This would store the builder for execution during run()
        // For now, we maintain the placeholder pattern but document the integration point
        self
    }

    /// Capture database schema using Phase 16 table introspection API
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if schema capture fails
    #[cfg(feature = "snapshots")]
    async fn capture_schema(&self, db: &dyn Database) -> Result<DatabaseSchema> {
        let mut schema = DatabaseSchema {
            tables: BTreeMap::new(),
        };

        // Use Phase 16 table introspection API to get schema information
        for table_name in &self.expected_tables {
            if let Some(table_info) = db.get_table_info(table_name).await? {
                // Convert Phase 16 TableInfo to our snapshot types
                let columns = table_info
                    .columns
                    .into_values()
                    .map(|col| ColumnInfo {
                        name: col.name,
                        data_type: format!("{:?}", col.data_type), // Convert DataType enum to string
                        nullable: col.nullable,
                        default_value: col.default_value.map(|v| format!("{v:?}")),
                        primary_key: col.is_primary_key,
                    })
                    .collect();

                let indexes = table_info
                    .indexes
                    .into_values()
                    .map(|idx| idx.name)
                    .collect();

                schema
                    .tables
                    .insert(table_name.clone(), TableSchema { columns, indexes });
            }
        }

        Ok(schema)
    }

    /// Auto-discover tables from migrations if `expected_tables` is empty
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if table discovery fails
    #[cfg(feature = "snapshots")]
    #[allow(unused, clippy::unused_async)] // Future enhancement
    async fn discover_tables_from_migrations(&self) -> Result<Vec<String>> {
        // TODO: Parse migration files in migrations_dir to find CREATE TABLE statements
        // For now, return empty vec - this would be implemented in a future enhancement
        Ok(vec![])
    }

    /// Capture data samples with type-aware conversion
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if data sampling fails
    #[cfg(feature = "snapshots")]
    async fn capture_data_samples(
        &self,
        db: &dyn Database,
    ) -> Result<std::collections::BTreeMap<String, Vec<serde_json::Value>>> {
        let mut samples = std::collections::BTreeMap::new();

        for (table, &count) in &self.data_samples {
            // Use Database query builder instead of raw SQL
            let query = db.select(table).limit(count);

            let rows = query.execute(db).await?;

            let sample_data: Vec<serde_json::Value> = rows
                .into_iter()
                .map(row_to_json) // Using our conversion function
                .collect();

            samples.insert(table.clone(), sample_data);
        }

        Ok(samples)
    }

    /// Create a test database using existing utilities
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if database creation fails
    #[cfg(feature = "snapshots")]
    async fn create_test_database(&self) -> Result<Box<dyn Database>> {
        log::debug!("Creating test database");
        // Use existing test_utils helper (SQLite in-memory)
        // This database persists for the entire test lifecycle
        let db = crate::create_empty_in_memory()
            .await
            .map_err(TestError::from)?;
        Ok(db)
    }

    /// Load migrations from directory that need to be applied
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if migration loading fails
    #[cfg(feature = "snapshots")]
    async fn load_migrations(&self) -> Result<Vec<Arc<dyn Migration<'static> + 'static>>> {
        if let Some(ref migrations_dir) = self.migrations_dir {
            if migrations_dir.exists() {
                log::debug!(
                    "Loading migrations from directory: {}",
                    migrations_dir.display()
                );

                let source = DirectoryMigrationSource::from_path(migrations_dir.clone());
                let migrations = source.migrations().await?;

                log::debug!("Loaded {} migrations from directory", migrations.len());
                return Ok(migrations);
            }

            log::debug!(
                "Migrations directory does not exist: {}",
                migrations_dir.display()
            );
        } else {
            log::debug!("No migrations directory configured");
        }

        Ok(vec![])
    }

    /// Get the sequence of already applied migrations from the database
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if querying fails
    #[cfg(feature = "snapshots")]
    async fn get_migration_sequence(&self, db: &dyn Database) -> Result<Vec<String>> {
        use switchy_schema::{migration::MigrationStatus, version::VersionTracker};

        let tracker = self
            .migrations_table_name
            .as_ref()
            .map_or_else(VersionTracker::new, |table_name| {
                VersionTracker::with_table_name(table_name.clone())
            });

        // Just call it directly - it handles missing table gracefully
        let ids = tracker
            .get_applied_migration_ids(db, MigrationStatus::Completed)
            .await
            .map_err(SnapshotError::Migration)?;

        log::debug!("Found {} applied migrations in database", ids.len());
        Ok(ids)
    }

    /// Run the snapshot test
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if test execution fails
    #[cfg(feature = "snapshots")]
    #[allow(clippy::cognitive_complexity)]
    pub async fn run(mut self) -> Result<()> {
        // Use provided database or create a new one
        let db = if let Some(db) = self.db.take() {
            db
        } else {
            self.create_test_database().await?
        };
        let db = &*db;

        // Load migrations from directory (these will be applied to the DB)
        let migrations_to_apply = self.load_migrations().await?;

        // Execute setup function if provided
        if let Some(setup_fn) = &self.setup_fn {
            log::debug!("run: executing setup function");
            setup_fn(db).await?;
        } else {
            log::debug!("run: no setup function provided");
        }

        // Execute migrations - fail fast on any error
        if migrations_to_apply.is_empty() {
            log::debug!("run: no new migrations to apply");
        } else {
            log::debug!("run: executing {} migrations", migrations_to_apply.len());
            let source = VecMigrationSource::new(migrations_to_apply.clone());
            let runner = MigrationRunner::new(Box::new(source));

            // Any migration error will propagate and fail the test
            runner.run(db).await?;
        }

        // Execute verification function if provided
        if let Some(verification_fn) = &self.verification_fn {
            log::debug!("run: executing verification function");
            verification_fn(db).await?;
        } else {
            log::debug!("run: no verification function provided");
        }

        // Capture results based on configuration
        let schema = if self.assert_schema {
            log::debug!("run: capturing schema");
            Some(self.capture_schema(db).await?)
        } else {
            log::debug!("run: no schema capture");
            None
        };

        // Get the sequence of already applied migrations AFTER running new ones
        let sequence = if self.assert_sequence {
            log::debug!("run: capturing migration sequence");
            self.get_migration_sequence(db).await?
        } else {
            log::debug!("run: no migration sequence capture");
            vec![]
        };
        log::debug!("run: migration sequence: {sequence:?}");

        let data_samples = if self.assert_data {
            log::debug!("run: capturing data samples");
            Some(self.capture_data_samples(db).await?)
        } else {
            log::debug!("run: no data samples capture");
            None
        };

        let snapshot = MigrationSnapshot {
            test_name: self.test_name.clone(),
            migration_sequence: sequence,
            schema,
            data_samples,
        };

        log::debug!("run: snapshot={snapshot:?}");

        // Apply redactions using insta's Settings with precise patterns
        let mut settings = Settings::clone_current();

        if self.redact_timestamps {
            // Precise timestamp patterns for different formats
            settings.add_filter(r"\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}", "[TIMESTAMP]");
            settings.add_filter(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}", "[TIMESTAMP]");
            settings.add_filter(r"\d{4}-\d{2}-\d{2}", "[DATE]");
        }

        if self.redact_auto_ids {
            // JSON-specific patterns for different ID fields
            settings.add_filter(r#""id": \d+"#, r#""id": "[ID]""#);
            settings.add_filter(r#""user_id": \d+"#, r#""user_id": "[USER_ID]""#);
            settings.add_filter(r#""post_id": \d+"#, r#""post_id": "[POST_ID]""#);
            settings.add_filter(r#""(\w+_id)": \d+"#, r#""$1": "[FK_ID]""#);
        }

        if self.redact_paths {
            // Unix and Windows path patterns
            settings.add_filter(r"/[\w/.-]+", "[PATH]");
            settings.add_filter(r"[A-Z]:\\[\w\\.-]+", "[PATH]");
        }

        settings.bind(|| {
            insta::assert_json_snapshot!(self.test_name, snapshot);
        });

        Ok(())
    }

    /// Run the snapshot test (non-snapshots version)
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if test execution fails
    #[cfg(not(feature = "snapshots"))]
    pub fn run(self) -> Result<()> {
        // Still minimal but uses configuration
        println!("Test: {}", self.test_name);
        println!("Migrations: {}", self.migrations_dir.display());
        println!(
            "Schema: {}, Sequence: {}",
            self.assert_schema, self.assert_sequence
        );

        Ok(())
    }
}

// Conversion functions from Phase 16 types to snapshot types
#[cfg(feature = "snapshots")]
#[allow(unused)]
fn table_info_to_schema(info: TableInfo) -> TableSchema {
    TableSchema {
        columns: info
            .columns
            .into_values()
            .map(db_column_info_to_column_info)
            .collect(),
        indexes: info.indexes.into_values().map(|idx| idx.name).collect(),
    }
}

#[cfg(feature = "snapshots")]
fn db_column_info_to_column_info(col: DbColumnInfo) -> ColumnInfo {
    ColumnInfo {
        name: col.name,
        data_type: format!("{:?}", col.data_type), // Convert DataType enum to string
        nullable: col.nullable,
        default_value: col.default_value.map(|v| format!("{v:?}")),
        primary_key: col.is_primary_key,
    }
}

// JSON Conversion functions for Row and DatabaseValue types
#[cfg(feature = "snapshots")]
#[allow(unused)]
fn row_to_json(row: Row) -> serde_json::Value {
    let map: serde_json::Map<String, serde_json::Value> = row
        .columns
        .into_iter()
        .map(|(k, v)| (k, database_value_to_json(v)))
        .collect();
    serde_json::Value::Object(map)
}

#[cfg(feature = "snapshots")]
#[allow(unused)]
fn database_value_to_json(value: DatabaseValue) -> serde_json::Value {
    match value {
        DatabaseValue::String(s) | DatabaseValue::StringOpt(Some(s)) => {
            serde_json::Value::String(s)
        }
        DatabaseValue::Bool(b) | DatabaseValue::BoolOpt(Some(b)) => serde_json::Value::Bool(b),
        DatabaseValue::Number(i) | DatabaseValue::NumberOpt(Some(i)) => {
            serde_json::Value::Number(i.into())
        }
        DatabaseValue::UNumber(u) | DatabaseValue::UNumberOpt(Some(u)) => {
            serde_json::Value::Number(u.into())
        }
        DatabaseValue::Real(f) | DatabaseValue::RealOpt(Some(f)) => serde_json::Number::from_f64(f)
            .map_or(serde_json::Value::Null, serde_json::Value::Number),
        DatabaseValue::Null
        | DatabaseValue::StringOpt(None)
        | DatabaseValue::BoolOpt(None)
        | DatabaseValue::NumberOpt(None)
        | DatabaseValue::UNumberOpt(None)
        | DatabaseValue::RealOpt(None) => serde_json::Value::Null,
        DatabaseValue::DateTime(dt) => serde_json::Value::String(dt.to_string()),
        DatabaseValue::NowAdd(s) => serde_json::Value::String(format!("NOW + {s}")),
        DatabaseValue::Now => serde_json::Value::String("NOW".to_string()),
    }
}
