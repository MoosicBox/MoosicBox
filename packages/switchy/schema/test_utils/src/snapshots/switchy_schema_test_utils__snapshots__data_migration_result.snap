---
source: packages/switchy/schema/test_utils/src/snapshots.rs
expression: snapshot
---
{
  "test_name": "data_migration_result",
  "migration_sequence": [
    "001_create_table",
    "002_add_column",
    "003_create_index"
  ],
  "schema": {
    "tables": {}
  },
  "data_samples": {
    "users": []
  }
}
